### PR TITLE
feat: add response received column

### PR DESCRIPTION
**Pull request by im-adithya**
This adds a response received column to the request events table so it's easy to check what events missed getting a response back.

(Also adds a migration so the previous request events are covered too!)

### CODE REVIEW

File: internal/nostr/models.go [hunk 1]
```diff
@@ -130,13 +130,14 @@ type OnReceiveEOSFunc func(ctx context.Context, subscription *Subscription)
...
-	Content        string
-	State          string
-	CreatedAt      time.Time
-	UpdatedAt      time.Time
+	ID               uint
+	SubscriptionId   *uint
+	NostrId          string    `validate:"required"`
+	Content          string
+	State            string
+	ResponseReceived bool
```
Comment (bumi) [inline internal/nostr/models.go:138]: maybe if you make it a datetime field then you know that it was received if it is filled
and you know the time of when it was received and could calculate the difference between the createat time and and the receivedat time

Comment (im-adithya) [reply internal/nostr/models.go:138]: Ah nice! That's a good idea!

File: migrations/202407171220_add_response_state_to_request_events.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import (
+	"github.com/go-gormigrate/gormigrate/v2"
+	"gorm.io/gorm"
+)
+
+// Add response_received_at column to request_events table
+var _202407171220_add_response_received_at_to_request_events = &gormigrate.Migration{
+	ID: "202407171220_add_response_received_at_to_request_events",
+	Migrate: func(tx *gorm.DB) error {
+		if err := tx.Exec("ALTER TABLE request_events ADD COLUMN response_received_at TIMESTAMP DEFAULT '0001-01-01 00:00:00+00'").Error; err != nil {
```
Comment (bumi) [inline migrations/202407171220_add_response_state_to_request_events.go:12]: I don't think a timestamp should have such a default, why did you add this? 

it can be nil. 

nil means no response received
and a timestamp indicates that and when a response is received

File: migrations/202407171220_add_response_state_to_request_events.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+			return err
+		}
+
+		// Update response_received_at if there is a corresponding row in response_events
+		if err := tx.Exec(`
+			UPDATE request_events re
+			SET response_received_at = (SELECT created_at FROM response_events WHERE request_id = re.id)
+			WHERE id IN (SELECT request_id FROM response_events WHERE request_id IS NOT NULL)
```
Comment (bumi) [inline migrations/202407171220_add_response_state_to_request_events.go:20]: subqueries alwasys look dangerous to me. 

and I am not sure how this works on a big dataset. you always have to wonder how does this work on millions of rows. what does this mean? does the table get locked during that time and thus the app becomes unusable because of that. 

I would say for the old data just set the response_received_at to the created_at column.

Comment (im-adithya) [reply migrations/202407171220_add_response_state_to_request_events.go:20]: > I would say for the old data just set the response_received_at to the created_at column.

But wouldn't that wrongly set it for all the request events which didn't actually receive a response? Are we okay with that?

> and I am not sure how this works on a big dataset. you always have to wonder how does this work on millions of rows. what does this mean? does the table get locked during that time and thus the app becomes unusable because of that.

Can I test this and verify on a sample db locally? Or you think this is not worth it?

File: migrations/202407171220_add_response_state_to_request_events.go [hunk 4]
```diff
@@ -0,0 +1,33 @@
+package migrations
+
+import (
+	"github.com/go-gormigrate/gormigrate/v2"
+	"gorm.io/gorm"
+)
+
+// Add response_received_at column to request_events table
+var _202407171220_add_response_received_at_to_request_events = &gormigrate.Migration{
```
[the remark this answers is not in the document]
Comment (im-adithya) [reply migrations/202407171220_add_response_state_to_request_events.go:9]: > keep that in sync with the file name.

Whoops, missed it thanks 🙌 

> and why is this migration deleted

Ah I already ran it locally so I had to change the filename to run the migration again for testing 😅

File: migrations/202407171220_add_response_state_to_request_events.go (hunk 2 above)
Comment (im-adithya) [reply migrations/202407171220_add_response_state_to_request_events.go]: I thought it would be easy to handle it in go if we wish to do something with these values later, but yeah it's not that much of a hassle and it's actually better while querying in metabase 👍

File: migrations/202407171220_add_response_state_to_request_events.go (hunk 3 above)
Comment (bumi) [reply migrations/202407171220_add_response_state_to_request_events.go]: > Can I test this and verify on a sample db locally? Or you think this is not worth it?

try to learn about it. you will not have the same data locally, so how to do that? even with the same data you will not have the same amount of requests coming in locally. 
what does the internet and the experience say? 
the main thing is that you have to anticipate such potential issues and then try to find answers.

File: migrations/202407171220_add_response_state_to_request_events.go (hunk 2 above)
Comment (bumi) [reply migrations/202407171220_add_response_state_to_request_events.go]: >  thought it would be easy to handle it in go if we wish to do something with these values later, 

what does that mean? 
this is in the DB and you have to think about correct DB values.

File: migrations/202407171220_add_response_state_to_request_events.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (bumi) [reply migrations/202407171220_add_response_state_to_request_events.go]: always review what you commit and if that makes sense. I assume in this commit you did not go through each change and each file when staging it to commit. (there is a reason also why git has this two step process)

and never do such hacks it will cause all this trouble. you're doing it wrong then, you have to adjust your setup and not the code. 
(at least until you really really know what you're doing and not making such mistakes and commits as you do here)

[the remark this answers is not in the document]
Comment (im-adithya) [reply migrations/202407171220_add_response_state_to_request_events.go]: > I assume in this commit you did not go through each change and each file when staging it to commit

No no, I am checking commits before staging one by one now! This genuinely got missed because it shows the file name as 202407131920_add_respon... in the sidebar, should've checked that as well, my bad

> and never do such hacks it will cause all this trouble. you're doing it wrong then, you have to adjust your setup and not the code

Got it!

File: internal/nostr/nostr.go [hunk 5]
```diff
@@ -794,6 +794,10 @@ func (svc *Service) handleResponseEvent(event *nostr.Event, subscription *Subscr
 		"request_event_id": subscription.RequestEvent.ID,
 		"wallet_pubkey":    svc.getWalletPubkey(subscription.Authors),
 	}).Info("Received response event")
+	if (subscription.RequestEvent != nil) {
```
Comment (bumi) [inline internal/nostr/nostr.go:797]: why is it here `RequestEvent` and then `RequestEventDB`?

Comment (im-adithya) [reply internal/nostr/nostr.go:797]: Opened #98 to tackle this separately

Comment (im-adithya) [reply internal/nostr/nostr.go:795]: Done
